src/cli.ts

```diff
@@ -37,44 +34,68 @@ const options = yargs()
       description: "Use char length as objective",
       type: "boolean",
     },
+    all: {
+      alias: "a",
+      description: "Get all variants",
+      type: "boolean",
+    },
+    debug: {
+      alias: "d",
+      description: "Print debug info, like history of applied plugins.",
+      type: "boolean",
+    },
   })
   .parseSync(process.argv.slice(2));
 
-const langs = options.lang === "all" ? languages : [findLang(options.lang)!];
+if (options.all === true && options.output !== undefined) {
+  throw new Error(
+    "All variants options is only allowed when the output file is not specified."
+  );
+}
+
+const langs =
+  options.lang === undefined ? languages : [findLang(options.lang)!];
 let input = options.input;
 if (!fs.existsSync(input)) input += ".polygolf";
 const code = fs.readFileSync(input, { encoding: "utf-8" });
-try {
-  const prog = parse(code);
-  const printingMultipleLangs =
-    langs.length > 1 && options.output === undefined;
-  for (const lang of langs) {
-    if (printingMultipleLangs) console.log(lang.name);
-    try {
-      const result = applyLanguage(
-        lang,
-        prog,
-        searchOptions("full", options.chars === true ? "chars" : "bytes")
+const printingMultipleLangs = langs.length > 1 && options.output === undefined;
+for (const result of compile(
+  code,
+  {
+    level: "full",
+    objective: options.chars === true ? "chars" : "bytes",
+    getAllVariants: options.all,
+  },
+  ...langs
+)) {
+  if (printingMultipleLangs) console.log(result.language);
+  if (typeof result.result === "string") {
+    if (options.output !== undefined) {
+      fs.mkdirSync(path.dirname(options.output), { recursive: true });
+      fs.writeFileSync(
+        options.output +
+          (langs.length > 1 || !options.output.includes(".")
+            ? "." + findLang(result.language)!.extension
+            : ""),
+        result.result
       );
-      if (options.output !== undefined) {
-        fs.mkdirSync(path.dirname(options.output), { recursive: true });
-        fs.writeFileSync(
-          options.output +
-            (langs.length > 1 || !options.output.includes(".")
-              ? "." + lang.extension
-              : ""),
-          result
-        );
-      } else {
-        console.log(result);
-      }
-    } catch (e) {
-      handleError(e);
+    } else {
+      console.log(result.result);
+    }
+    if (result.warnings.length > 0) {
+      console.log("Warnings:");
+      console.log(result.warnings.map((x) => x.message).join("\n"));
```

Comment:
Should warnings be printed only if `options.debug === true`?
Should these go to stderr? (same with the other debugs down below).

Reply:
My idea was that the debug option is for debugging Polygolf, not really for debugging the Polygolf programs. But I guess it also depends on how the warnings will actually be used, there are currently none being generated.

Reply:
Maybe, instead of a boolean flag for debug, we could have a verbosity scale.